File: rxs3799_lab01.py
# Name: Riyan Shrestha
# ID: 1002223799
# Language Version: Python 3.12.1
# OS: Ubuntu 24.04

import os
import sys

def calculate_size(path= str) -> int:

    total_size = 0
    
    try:
        # Get list of all items in the current directory
        items = os.listdir(path)
        
        # Iterate through each item in the directory
        for item in items:
            item_path = os.path.join(path, item)
            
            try:
                # Check if the item is a directory
                if os.path.isdir(item_path):
                    # Recursively call this function for subdirectories
                    total_size += calculate_size(item_path)
                elif os.path.isfile(item_path):
                    # Add the file size to the total
                    total_size += os.path.getsize(item_path)
            except (OSError, IOError):
                # Skip files/directories that cannot be accessed due to permissions
                continue
                
    except (OSError, IOError):
        # Handle case where directory cannot be accessed
        return 0
    
    return total_size
# Entry point of the program
if __name__ == "__main__":
    start_path = sys.argv[1] if len(sys.argv) > 1 else "."
    size_bytes = calculate_size(start_path)
    print(size_bytes)

File: rxs3799_lab01.cpp
//Name: Riyan Shrestha
//ID: 1002223799
//Language Version: C++ (GCC 13.3.0)
//OS: Ubuntu 24.04 (Omega server)

#include <iostream>
#include <filesystem>

using namespace std;
namespace fs = filesystem;

uintmax_t calculate_size(const fs::path& path) {
    uintmax_t total = 0;
    error_code ec;  // Used to handle errors without exceptions

    // Iterate through all entries in the current directory
    for (const auto& entry : fs::directory_iterator(path, ec)) {
        if (ec) {
            continue; // Skip entries thta cannot be read
        }

        // Check if the current entry is a directory
        if (entry.is_directory(ec)) {
            // Recurse into the subdirectory
            total += calculate_size(entry.path());
        } else if (entry.is_regular_file(ec)) {
            // Add size of a normal file
            auto sz = entry.file_size(ec);
            if (!ec) {
                total += sz;
            }
        }
    }

    return total;
}

int main(int argc, char* argv[]) {
    try {
        fs::path start_dir = (argc > 1) ? fs::path(argv[1]) : fs::path(".");
        // Use command line argument for the starting directory or use the current directory if not provided 
        uintmax_t bytes = calculate_size(start_dir);
        cout << bytes << endl;
    } catch (const fs::filesystem_error& ex) {
        cerr << "Filesystem error: " << ex.what() << endl;
        return 1;
    }
    return 0;
}
